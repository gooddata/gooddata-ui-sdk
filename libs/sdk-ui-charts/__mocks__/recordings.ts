// (C) 2020 GoodData Corporation

import { DataViewFacade } from "@gooddata/sdk-ui";
import { ScenarioRecording, recordedDataView } from "@gooddata/sdk-backend-mockingbird";

export function recordedDataFacade(rec: ScenarioRecording): DataViewFacade {
    return DataViewFacade.for(recordedDataView(rec));
}
