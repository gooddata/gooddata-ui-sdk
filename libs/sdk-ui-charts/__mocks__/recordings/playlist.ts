// (C) 2019-2019 GoodData Corporation

import { createRequire } from "module";
const require = createRequire(import.meta.url);

// THIS FILE WAS AUTO-GENERATED ON 2019-11-05T10:50:00.366Z; DO NOT EDIT; LOOK INTO 'mock-handling' TOOLING

export const AreaChartWith3MetricsAndViewByAttribute = {
    definition: require("./execution/area_chart_with_3_metrics_and_view_by_attribute/definition.json"),
    response: require("./execution/area_chart_with_3_metrics_and_view_by_attribute/response.json"),
    result: require("./execution/area_chart_with_3_metrics_and_view_by_attribute/result.json"),
};

export const AreaChartWithMeasureViewByAndStackBy = {
    definition: require("./execution/area_chart_with_measure_view_by_and_stack_by/definition.json"),
    response: require("./execution/area_chart_with_measure_view_by_and_stack_by/response.json"),
    result: require("./execution/area_chart_with_measure_view_by_and_stack_by/result.json"),
};

export const BarChartWith3MetricsAndViewByAttribute = {
    definition: require("./execution/bar_chart_with_3_metrics_and_view_by_attribute/definition.json"),
    response: require("./execution/bar_chart_with_3_metrics_and_view_by_attribute/response.json"),
    result: require("./execution/bar_chart_with_3_metrics_and_view_by_attribute/result.json"),
};

export const BarChartWith3MetricsAndViewByAttributeFunformat = {
    definition: require("./execution/bar_chart_with_3_metrics_and_view_by_attribute_funformat/definition.json"),
    response: require("./execution/bar_chart_with_3_metrics_and_view_by_attribute_funformat/response.json"),
    result: require("./execution/bar_chart_with_3_metrics_and_view_by_attribute_funformat/result.json"),
};

export const BarChartWith3MetricsAndViewByAttributePercInFormat = {
    definition: require("./execution/bar_chart_with_3_metrics_and_view_by_attribute_perc_in_format/definition.json"),
    response: require("./execution/bar_chart_with_3_metrics_and_view_by_attribute_perc_in_format/response.json"),
    result: require("./execution/bar_chart_with_3_metrics_and_view_by_attribute_perc_in_format/result.json"),
};

export const BarChartWith4MetricsAndViewByTwoAttributes = {
    definition: require("./execution/bar_chart_with_4_metrics_and_view_by_two_attributes/definition.json"),
    response: require("./execution/bar_chart_with_4_metrics_and_view_by_two_attributes/response.json"),
    result: require("./execution/bar_chart_with_4_metrics_and_view_by_two_attributes/result.json"),
};

export const BarChartWithPopMeasureAndViewByAttribute = {
    definition: require("./execution/bar_chart_with_pop_measure_and_view_by_attribute/definition.json"),
    response: require("./execution/bar_chart_with_pop_measure_and_view_by_attribute/response.json"),
    result: require("./execution/bar_chart_with_pop_measure_and_view_by_attribute/result.json"),
};

export const BarChartWithPopMeasureAndViewByAttributeX6 = {
    definition: require("./execution/bar_chart_with_pop_measure_and_view_by_attribute_x6/definition.json"),
    response: require("./execution/bar_chart_with_pop_measure_and_view_by_attribute_x6/response.json"),
    result: require("./execution/bar_chart_with_pop_measure_and_view_by_attribute_x6/result.json"),
};

export const BarChartWithPreviousPeriodMeasure = {
    definition: require("./execution/bar_chart_with_previous_period_measure/definition.json"),
    response: require("./execution/bar_chart_with_previous_period_measure/response.json"),
    result: require("./execution/bar_chart_with_previous_period_measure/result.json"),
};

export const BarChartWithPreviousPeriodMeasureX6 = {
    definition: require("./execution/bar_chart_with_previous_period_measure_x6/definition.json"),
    response: require("./execution/bar_chart_with_previous_period_measure_x6/response.json"),
    result: require("./execution/bar_chart_with_previous_period_measure_x6/result.json"),
};

export const BarChartWithSingleMeasureAndNoAttributes = {
    definition: require("./execution/bar_chart_with_single_measure_and_no_attributes/definition.json"),
    response: require("./execution/bar_chart_with_single_measure_and_no_attributes/response.json"),
    result: require("./execution/bar_chart_with_single_measure_and_no_attributes/result.json"),
};

export const BarChartWithStackByAndViewByAttributes = {
    definition: require("./execution/bar_chart_with_stack_by_and_view_by_attributes/definition.json"),
    response: require("./execution/bar_chart_with_stack_by_and_view_by_attributes/response.json"),
    result: require("./execution/bar_chart_with_stack_by_and_view_by_attributes/result.json"),
};

export const BarChartWithViewByAttribute = {
    definition: require("./execution/bar_chart_with_view_by_attribute/definition.json"),
    response: require("./execution/bar_chart_with_view_by_attribute/response.json"),
    result: require("./execution/bar_chart_with_view_by_attribute/result.json"),
};

export const BubbleChartWith1Metric = {
    definition: require("./execution/bubble_chart_with_1_metric/definition.json"),
    response: require("./execution/bubble_chart_with_1_metric/response.json"),
    result: require("./execution/bubble_chart_with_1_metric/result.json"),
};

export const BubbleChartWith2MetricsAndAttributeNoPrimaries = {
    definition: require("./execution/bubble_chart_with_2_metrics_and_attribute_no_primaries/definition.json"),
    response: require("./execution/bubble_chart_with_2_metrics_and_attribute_no_primaries/response.json"),
    result: require("./execution/bubble_chart_with_2_metrics_and_attribute_no_primaries/result.json"),
};

export const BubbleChartWith3MetricsAndAttribute = {
    definition: require("./execution/bubble_chart_with_3_metrics_and_attribute/definition.json"),
    response: require("./execution/bubble_chart_with_3_metrics_and_attribute/response.json"),
    result: require("./execution/bubble_chart_with_3_metrics_and_attribute/result.json"),
};

export const BubbleChartWith3MetricsAndAttributeNullsInData = {
    definition: require("./execution/bubble_chart_with_3_metrics_and_attribute_nulls_in_data/definition.json"),
    response: require("./execution/bubble_chart_with_3_metrics_and_attribute_nulls_in_data/response.json"),
    result: require("./execution/bubble_chart_with_3_metrics_and_attribute_nulls_in_data/result.json"),
};

export const ComboChartWithTwoMeasuresViewByAttribute = {
    definition: require("./execution/combo_chart_with_two_measures_view_by_attribute/definition.json"),
    response: require("./execution/combo_chart_with_two_measures_view_by_attribute/response.json"),
    result: require("./execution/combo_chart_with_two_measures_view_by_attribute/result.json"),
};

export const ComboChartWithTwoMeasuresViewByAttributeNoBuckets = {
    definition: require("./execution/combo_chart_with_two_measures_view_by_attribute_no_buckets/definition.json"),
    response: require("./execution/combo_chart_with_two_measures_view_by_attribute_no_buckets/response.json"),
    result: require("./execution/combo_chart_with_two_measures_view_by_attribute_no_buckets/result.json"),
};

export const ComboChartWithTwoMeasuresViewByAttributePercformat = {
    definition: require("./execution/combo_chart_with_two_measures_view_by_attribute_percformat/definition.json"),
    response: require("./execution/combo_chart_with_two_measures_view_by_attribute_percformat/response.json"),
    result: require("./execution/combo_chart_with_two_measures_view_by_attribute_percformat/result.json"),
};

export const HeadlineWithOneMeasure = {
    definition: require("./execution/headline_with_one_measure/definition.json"),
    response: require("./execution/headline_with_one_measure/response.json"),
    result: require("./execution/headline_with_one_measure/result.json"),
};

export const HeadlineWithOneMeasureWithIdentifier = {
    definition: require("./execution/headline_with_one_measure_with_identifier/definition.json"),
    response: require("./execution/headline_with_one_measure_with_identifier/response.json"),
    result: require("./execution/headline_with_one_measure_with_identifier/result.json"),
};

export const HeadlineWithTwoMeasures = {
    definition: require("./execution/headline_with_two_measures/definition.json"),
    response: require("./execution/headline_with_two_measures/response.json"),
    result: require("./execution/headline_with_two_measures/result.json"),
};

export const HeadlineWithTwoMeasuresBothEmpty = {
    definition: require("./execution/headline_with_two_measures_both_empty/definition.json"),
    response: require("./execution/headline_with_two_measures_both_empty/response.json"),
    result: require("./execution/headline_with_two_measures_both_empty/result.json"),
};

export const HeadlineWithTwoMeasuresBothSame = {
    definition: require("./execution/headline_with_two_measures_both_same/definition.json"),
    response: require("./execution/headline_with_two_measures_both_same/response.json"),
    result: require("./execution/headline_with_two_measures_both_same/result.json"),
};

export const HeadlineWithTwoMeasuresBothZero = {
    definition: require("./execution/headline_with_two_measures_both_zero/definition.json"),
    response: require("./execution/headline_with_two_measures_both_zero/response.json"),
    result: require("./execution/headline_with_two_measures_both_zero/result.json"),
};

export const HeadlineWithTwoMeasuresFirstEmpty = {
    definition: require("./execution/headline_with_two_measures_first_empty/definition.json"),
    response: require("./execution/headline_with_two_measures_first_empty/response.json"),
    result: require("./execution/headline_with_two_measures_first_empty/result.json"),
};

export const HeadlineWithTwoMeasuresFirstZero = {
    definition: require("./execution/headline_with_two_measures_first_zero/definition.json"),
    response: require("./execution/headline_with_two_measures_first_zero/response.json"),
    result: require("./execution/headline_with_two_measures_first_zero/result.json"),
};

export const HeadlineWithTwoMeasuresSecondEmpty = {
    definition: require("./execution/headline_with_two_measures_second_empty/definition.json"),
    response: require("./execution/headline_with_two_measures_second_empty/response.json"),
    result: require("./execution/headline_with_two_measures_second_empty/result.json"),
};

export const HeadlineWithTwoMeasuresSecondZero = {
    definition: require("./execution/headline_with_two_measures_second_zero/definition.json"),
    response: require("./execution/headline_with_two_measures_second_zero/response.json"),
    result: require("./execution/headline_with_two_measures_second_zero/result.json"),
};

export const HeadlineWithTwoMeasuresWithIdentifier = {
    definition: require("./execution/headline_with_two_measures_with_identifier/definition.json"),
    response: require("./execution/headline_with_two_measures_with_identifier/response.json"),
    result: require("./execution/headline_with_two_measures_with_identifier/result.json"),
};

export const HeatMapWithEmptyCells = {
    definition: require("./execution/heat_map_with_empty_cells/definition.json"),
    response: require("./execution/heat_map_with_empty_cells/response.json"),
    result: require("./execution/heat_map_with_empty_cells/result.json"),
};

export const HeatMapWithMetricRowColumn = {
    definition: require("./execution/heat_map_with_metric_row_column/definition.json"),
    response: require("./execution/heat_map_with_metric_row_column/response.json"),
    result: require("./execution/heat_map_with_metric_row_column/result.json"),
};

export const PieChartWithMetricsOnly = {
    definition: require("./execution/pie_chart_with_metrics_only/definition.json"),
    response: require("./execution/pie_chart_with_metrics_only/response.json"),
    result: require("./execution/pie_chart_with_metrics_only/result.json"),
};

export const PieChartWithMetricsOnlyFundata = {
    definition: require("./execution/pie_chart_with_metrics_only_fundata/definition.json"),
    response: require("./execution/pie_chart_with_metrics_only_fundata/response.json"),
    result: require("./execution/pie_chart_with_metrics_only_fundata/result.json"),
};

export const ScatterPlotWith2MetricsAndAttributeNullsInData = {
    definition: require("./execution/scatter_plot_with_2_metrics_and_attribute_nulls_in_data/definition.json"),
    response: require("./execution/scatter_plot_with_2_metrics_and_attribute_nulls_in_data/response.json"),
    result: require("./execution/scatter_plot_with_2_metrics_and_attribute_nulls_in_data/result.json"),
};

export const ScatterPlotWith2MetricsAndAttributeWithPrimary = {
    definition: require("./execution/scatter_plot_with_2_metrics_and_attribute_with_primary/definition.json"),
    response: require("./execution/scatter_plot_with_2_metrics_and_attribute_with_primary/response.json"),
    result: require("./execution/scatter_plot_with_2_metrics_and_attribute_with_primary/result.json"),
};

export const TreemapWithMetricAndStackByAttribute = {
    definition: require("./execution/treemap_with_metric_and_stack_by_attribute/definition.json"),
    response: require("./execution/treemap_with_metric_and_stack_by_attribute/response.json"),
    result: require("./execution/treemap_with_metric_and_stack_by_attribute/result.json"),
};

export const TreemapWithMetricAndViewByAttribute = {
    definition: require("./execution/treemap_with_metric_and_view_by_attribute/definition.json"),
    response: require("./execution/treemap_with_metric_and_view_by_attribute/response.json"),
    result: require("./execution/treemap_with_metric_and_view_by_attribute/result.json"),
};

export const TreemapWithMetricViewByAndStackByAttribute = {
    definition: require("./execution/treemap_with_metric_view_by_and_stack_by_attribute/definition.json"),
    response: require("./execution/treemap_with_metric_view_by_and_stack_by_attribute/response.json"),
    result: require("./execution/treemap_with_metric_view_by_and_stack_by_attribute/result.json"),
};

export const TreemapWithTwoMetricsAndStackByAttribute = {
    definition: require("./execution/treemap_with_two_metrics_and_stack_by_attribute/definition.json"),
    response: require("./execution/treemap_with_two_metrics_and_stack_by_attribute/response.json"),
    result: require("./execution/treemap_with_two_metrics_and_stack_by_attribute/result.json"),
};

// initialize recorded backend with this
export const MasterIndex = {
    testWorkspace: {
        execution: {
            fp_0b599ac8ef275e712d231f304d9e29d0: AreaChartWithMeasureViewByAndStackBy,
            fp_45eabf83c34eaa23f347b56a5858d57f: ComboChartWithTwoMeasuresViewByAttributeNoBuckets,
            fp_baa24e54ab2909e7a7ae4f5f6d21ffaa: BarChartWith3MetricsAndViewByAttributeFunformat,
            fp_d04ba7363a1ce9162b18fe0ddc005a07: BarChartWith3MetricsAndViewByAttributePercInFormat,
            fp_3f95917db9f15eb872d904631fc3edcd: BarChartWith4MetricsAndViewByTwoAttributes,
            fp_f567a925d82dd41eff9780eb4ca934d2: BarChartWithPopMeasureAndViewByAttribute,
            fp_62bcd4f8af06c91324c3af201560ac9c: BarChartWithPopMeasureAndViewByAttributeX6,
            fp_ac28e4e88b5011dd98c8c92f970f3910: BarChartWithPreviousPeriodMeasure,
            fp_b836c933d95c1aef2d07f950f38b645d: BarChartWithPreviousPeriodMeasureX6,
            fp_7ebc96ae6d02197155fcd63a41f870b3: BarChartWithSingleMeasureAndNoAttributes,
            fp_1b740fffd1df15049b4c5771cf293327: BarChartWithStackByAndViewByAttributes,
            fp_8d33743b38a06a7af0a1a71e3c573f8b: BarChartWithViewByAttribute,
            fp_4b146f4139daf56999dc16e76d478de9: BubbleChartWith1Metric,
            fp_14f9897772326f7e811f4aa9a20d871d: BubbleChartWith2MetricsAndAttributeNoPrimaries,
            fp_bcc9d43c8427eeb41548d866b1630374: BubbleChartWith3MetricsAndAttributeNullsInData,
            fp_63a093ee40d9e5d13e14cd186ced8376: ComboChartWithTwoMeasuresViewByAttributePercformat,
            fp_e2b71aa2be7ba9d7c46ee9f153cce9f2: HeadlineWithOneMeasure,
            fp_4f043c79dbefafee7e26f3f77d945b86: HeadlineWithOneMeasureWithIdentifier,
            fp_d53d09f0ed2ab5f3ef73b73ca749e4f7: HeadlineWithTwoMeasuresSecondZero,
            fp_4bf5193c9d1f6560bc5c43b0b2b99ea1: HeadlineWithTwoMeasuresWithIdentifier,
            fp_7ba00bd46fb37f68fbb549e997f349dd: HeatMapWithEmptyCells,
            fp_16dd93e14c96b0317d74f5e0e37792ff: HeatMapWithMetricRowColumn,
            fp_551204e9d10f29532c64d0aad6d9b19b: PieChartWithMetricsOnlyFundata,
            fp_f28b82a9ed53ff7655a633dca593e1fb: ScatterPlotWith2MetricsAndAttributeWithPrimary,
            fp_539276f8275d76bce58f585086fd2cf4: TreemapWithMetricViewByAndStackByAttribute,
            fp_d81a1d01654699ff102ae5be3839bcde: TreemapWithTwoMetricsAndStackByAttribute,
        },
    },
};
