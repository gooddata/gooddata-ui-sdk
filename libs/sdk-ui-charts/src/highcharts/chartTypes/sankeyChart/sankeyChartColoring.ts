// (C) 2023 GoodData Corporation
import uniqBy from "lodash/uniqBy.js";
import {
    ColorStrategy,
    getColorFromMapping,
    IColorMapping,
    ICreateColorAssignmentReturnValue,
    isValidMappedColor,
} from "@gooddata/sdk-ui-vis-commons";
import { IColor, IColorPalette } from "@gooddata/sdk-model";
import { DataViewFacade } from "@gooddata/sdk-ui";
import compact from "lodash/compact.js";

export class SankeyChartColorStrategy extends ColorStrategy {
    protected createColorAssignment(
        colorPalette: IColorPalette,
        colorMapping: IColorMapping[],
        fromAttribute: any,
        toAttribute: any,
        dv: DataViewFacade,
    ): ICreateColorAssignmentReturnValue {
        const attributeItems = compact([...(fromAttribute?.items ?? []), ...(toAttribute?.items ?? [])]);
        const uniqAttributeItems = uniqBy(attributeItems, "attributeHeaderItem.uri");

        const colorAssignment = uniqAttributeItems.map((headerItem, index) => {
            const mappedColor = getColorFromMapping(headerItem, colorMapping, dv);
            const color: IColor =
                mappedColor && isValidMappedColor(mappedColor, colorPalette)
                    ? mappedColor
                    : {
                          type: "guid",
                          value: colorPalette[index % colorPalette.length].guid,
                      };
            return {
                headerItem,
                color,
            };
        });
        return {
            fullColorAssignment: colorAssignment,
        };
    }
}
