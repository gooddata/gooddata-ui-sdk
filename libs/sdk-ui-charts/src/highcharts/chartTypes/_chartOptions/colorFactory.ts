// (C) 2007-2024 GoodData Corporation
import { IColorPalette, ITheme } from "@gooddata/sdk-model";
import { DataViewFacade, DefaultColorPalette, VisualizationTypes } from "@gooddata/sdk-ui";
import { IColorMapping } from "../../../interfaces/index.js";
import {
    isBubbleChart,
    isBulletChart,
    isHeatmap,
    isOneOfTypes,
    isSankeyOrDependencyWheel,
    isScatterPlot,
    isTreemap,
    isWaterfall,
} from "../_util/common.js";
import BulletChartColorStrategy from "../bulletChart/bulletChartColoring.js";
import { MeasureColorStrategy } from "../_chartColoring/measure.js";
import { AttributeColorStrategy, IColorStrategy } from "@gooddata/sdk-ui-vis-commons";
import { HeatmapColorStrategy } from "../heatmap/heatmapColoring.js";
import { TreemapColorStrategy } from "../treemap/treemapColoring.js";
import { BubbleChartColorStrategy } from "../bubbleChart/bubbleChartColoring.js";
import { ScatterPlotColorStrategy } from "../scatterPlot/scatterPlotColoring.js";
import { SankeyChartColorStrategy } from "../sankeyChart/sankeyChartColoring.js";
import { WaterfallChartColorStrategy } from "../waterfallChart/waterfallChartColoring.js";

const attributeChartSupportedTypes = [
    VisualizationTypes.PIE,
    VisualizationTypes.DONUT,
    VisualizationTypes.FUNNEL,
    VisualizationTypes.PYRAMID,
    VisualizationTypes.SCATTER,
    VisualizationTypes.BUBBLE,
    VisualizationTypes.SANKEY,
    VisualizationTypes.DEPENDENCY_WHEEL,
];

function isAttributeColorPalette(type: string, dv: DataViewFacade, stackByAttribute: any) {
    const attributeChartSupported = isOneOfTypes(type, attributeChartSupportedTypes);
    return stackByAttribute || (attributeChartSupported && dv.def().hasAttributes());
}

/**
 * @internal
 */
export class ColorFactory {
    public static getColorStrategy(
        colorPalette: IColorPalette = DefaultColorPalette,
        colorMapping: IColorMapping[],
        // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
        viewByAttribute: any,
        // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
        viewByParentAttribute: any,
        // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
        stackByAttribute: any,
        dv: DataViewFacade,
        type: string,
        theme?: ITheme,
    ): IColorStrategy {
        if (isHeatmap(type)) {
            return new HeatmapColorStrategy(
                colorPalette,
                colorMapping,
                viewByAttribute,
                stackByAttribute,
                dv,
                theme,
            );
        }

        if (isTreemap(type)) {
            return new TreemapColorStrategy(
                colorPalette,
                colorMapping,
                viewByAttribute,
                stackByAttribute,
                dv,
                theme,
            );
        }

        if (isScatterPlot(type)) {
            return new ScatterPlotColorStrategy(
                colorPalette,
                colorMapping,
                viewByAttribute,
                stackByAttribute,
                dv,
                theme,
            );
        }

        if (isBubbleChart(type)) {
            return new BubbleChartColorStrategy(
                colorPalette,
                colorMapping,
                viewByAttribute,
                stackByAttribute,
                dv,
                theme,
            );
        }

        if (isBulletChart(type)) {
            return new BulletChartColorStrategy(
                colorPalette,
                colorMapping,
                viewByAttribute,
                stackByAttribute,
                dv,
                theme,
            );
        }

        if (isSankeyOrDependencyWheel(type)) {
            return new SankeyChartColorStrategy(
                colorPalette,
                colorMapping,
                viewByParentAttribute,
                viewByAttribute,
                dv,
                theme,
            );
        }

        if (isWaterfall(type)) {
            return new WaterfallChartColorStrategy(
                colorPalette,
                colorMapping,
                viewByParentAttribute,
                viewByAttribute,
                dv,
            );
        }

        if (isAttributeColorPalette(type, dv, stackByAttribute)) {
            return new AttributeColorStrategy(
                colorPalette,
                colorMapping,
                viewByAttribute,
                stackByAttribute,
                dv,
                theme,
            );
        }

        return new MeasureColorStrategy(
            colorPalette,
            colorMapping,
            viewByAttribute,
            stackByAttribute,
            dv,
            theme,
        );
    }
}
