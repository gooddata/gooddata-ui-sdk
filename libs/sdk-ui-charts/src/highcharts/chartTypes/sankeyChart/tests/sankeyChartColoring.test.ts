// (C) 2023 GoodData Corporation
import { getMVSForViewByTwoAttributes } from "../../_util/test/helper.js";
import { IColorStrategy } from "@gooddata/sdk-ui-vis-commons";
import { ColorFactory } from "../../_chartOptions/colorFactory.js";
import { SankeyChartColorStrategy } from "../sankeyChartColoring.js";
import { recordedDataFacade } from "../../../../../__mocks__/recordings.js";
import { CUSTOM_COLOR_PALETTE } from "../../_util/test/colorPalette.fixture.js";
import { CHART_TYPE, COLOR_MAPPINGS, RECORDS_WITHOUT_EMPTY_ATTRIBUTE } from "./sankeyChart.fixture.js";
import { describe, it, expect } from "vitest";

describe("SankeyChartColorStrategy", () => {
    it.each(RECORDS_WITHOUT_EMPTY_ATTRIBUTE)(
        "should return SankeyChartColorStrategy strategy in cases %s",
        (_title, record) => {
            const dv = recordedDataFacade(record);
            const { viewByAttribute, viewByParentAttribute } = getMVSForViewByTwoAttributes(dv);

            const colorStrategy: IColorStrategy = ColorFactory.getColorStrategy(
                undefined,
                undefined,
                viewByAttribute,
                viewByParentAttribute,
                null,
                dv,
                CHART_TYPE,
            );

            expect(colorStrategy).toBeInstanceOf(SankeyChartColorStrategy);
        },
    );

    it.each(RECORDS_WITHOUT_EMPTY_ATTRIBUTE)(
        "should return correct colors for each attribute header items in cases %s",
        (_title, record) => {
            const dv = recordedDataFacade(record);
            const { viewByAttribute, viewByParentAttribute } = getMVSForViewByTwoAttributes(dv);

            const colorStrategy: IColorStrategy = ColorFactory.getColorStrategy(
                CUSTOM_COLOR_PALETTE,
                undefined,
                viewByAttribute,
                viewByParentAttribute,
                null,
                dv,
                CHART_TYPE,
            );

            expect(colorStrategy.getColorAssignment()).toMatchSnapshot();
        },
    );

    it.each(RECORDS_WITHOUT_EMPTY_ATTRIBUTE)(
        "should return correct colors for each attribute header items when color mapping in cases %s",
        (_title, record) => {
            const dv = recordedDataFacade(record);
            const { viewByAttribute, viewByParentAttribute } = getMVSForViewByTwoAttributes(dv);

            const colorStrategy: IColorStrategy = ColorFactory.getColorStrategy(
                CUSTOM_COLOR_PALETTE,
                COLOR_MAPPINGS,
                viewByAttribute,
                viewByParentAttribute,
                null,
                dv,
                CHART_TYPE,
            );

            expect(colorStrategy.getColorAssignment()).toMatchSnapshot();
        },
    );
});
