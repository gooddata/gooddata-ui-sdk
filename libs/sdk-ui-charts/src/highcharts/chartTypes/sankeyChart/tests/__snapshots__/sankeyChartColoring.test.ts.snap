// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SankeyChartColorStrategy > should return correct colors for each attribute header items in cases 1 measure and 1 attribute_from 1`] = `
[
  {
    "color": {
      "type": "guid",
      "value": "01",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "CompuSci",
        "uri": "CompuSci",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "02",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "Educationly",
        "uri": "Educationly",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "03",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "Explorer",
        "uri": "Explorer",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "04",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "Grammar Plus",
        "uri": "Grammar Plus",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "05",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "PhoenixSoft",
        "uri": "PhoenixSoft",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "01",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "WonderKid",
        "uri": "WonderKid",
      },
    },
  },
]
`;

exports[`SankeyChartColorStrategy > should return correct colors for each attribute header items in cases 1 measure and 1 attribute_to 1`] = `
[
  {
    "color": {
      "type": "guid",
      "value": "01",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "East Coast",
        "uri": "East Coast",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "02",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "West Coast",
        "uri": "West Coast",
      },
    },
  },
]
`;

exports[`SankeyChartColorStrategy > should return correct colors for each attribute header items in cases 1 measure and 2 attributes 1`] = `
[
  {
    "color": {
      "type": "guid",
      "value": "01",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "CompuSci",
        "uri": "CompuSci",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "02",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "Educationly",
        "uri": "Educationly",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "03",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "Explorer",
        "uri": "Explorer",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "04",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "Grammar Plus",
        "uri": "Grammar Plus",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "05",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "PhoenixSoft",
        "uri": "PhoenixSoft",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "01",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "WonderKid",
        "uri": "WonderKid",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "02",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "East Coast",
        "uri": "East Coast",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "03",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "West Coast",
        "uri": "West Coast",
      },
    },
  },
]
`;

exports[`SankeyChartColorStrategy > should return correct colors for each attribute header items when color mapping in cases 1 measure and 1 attribute_from 1`] = `
[
  {
    "color": {
      "type": "guid",
      "value": "01",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "CompuSci",
        "uri": "CompuSci",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "02",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "Educationly",
        "uri": "Educationly",
      },
    },
  },
  {
    "color": {
      "type": "rgb",
      "value": {
        "b": 0,
        "g": 0,
        "r": 255,
      },
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "Explorer",
        "uri": "Explorer",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "04",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "Grammar Plus",
        "uri": "Grammar Plus",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "05",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "PhoenixSoft",
        "uri": "PhoenixSoft",
      },
    },
  },
  {
    "color": {
      "type": "rgb",
      "value": {
        "b": 0,
        "g": 0,
        "r": 0,
      },
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "WonderKid",
        "uri": "WonderKid",
      },
    },
  },
]
`;

exports[`SankeyChartColorStrategy > should return correct colors for each attribute header items when color mapping in cases 1 measure and 1 attribute_to 1`] = `
[
  {
    "color": {
      "type": "guid",
      "value": "01",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "East Coast",
        "uri": "East Coast",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "02",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "West Coast",
        "uri": "West Coast",
      },
    },
  },
]
`;

exports[`SankeyChartColorStrategy > should return correct colors for each attribute header items when color mapping in cases 1 measure and 2 attributes 1`] = `
[
  {
    "color": {
      "type": "guid",
      "value": "01",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "CompuSci",
        "uri": "CompuSci",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "02",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "Educationly",
        "uri": "Educationly",
      },
    },
  },
  {
    "color": {
      "type": "rgb",
      "value": {
        "b": 0,
        "g": 0,
        "r": 255,
      },
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "Explorer",
        "uri": "Explorer",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "04",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "Grammar Plus",
        "uri": "Grammar Plus",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "05",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "PhoenixSoft",
        "uri": "PhoenixSoft",
      },
    },
  },
  {
    "color": {
      "type": "rgb",
      "value": {
        "b": 0,
        "g": 0,
        "r": 0,
      },
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "WonderKid",
        "uri": "WonderKid",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "02",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "East Coast",
        "uri": "East Coast",
      },
    },
  },
  {
    "color": {
      "type": "guid",
      "value": "03",
    },
    "headerItem": {
      "attributeHeaderItem": {
        "name": "West Coast",
        "uri": "West Coast",
      },
    },
  },
]
`;
